File: internal/commands/grouproles.go
// Package commands provides CLI commands.
package commands

import (
	"context"
	"fmt"
	"net/url"

	"github.com/spf13/cobra"

	"github.com/jacebenson/jsn/internal/appctx"
	"github.com/jacebenson/jsn/internal/output"
)

// NewGroupRolesCmd creates the grouproles command group.
func NewGroupRolesCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:     "grouproles",
		Aliases: []string{"grouprole", "grole", "gr"},
		Short:   "Manage group roles",
		Long: `Manage role assignments to ServiceNow groups.

Examples:
  # Add a role to a group
  jsn grouproles add --group "IT Support" --role "itil"

  # Remove a role from a group
  jsn grouproles remove --group "IT Support" --role "itil"

  # List roles of a group
  jsn grouproles list --group "IT Support"`,
	}

	cmd.AddCommand(
		newGroupRolesAddCmd(),
		newGroupRolesRemoveCmd(),
		newGroupRolesListCmd(),
	)

	return cmd
}

func newGroupRolesAddCmd() *cobra.Command {
	var (
		groupName string
		roleName  string
	)

	cmd := &cobra.Command{
		Use:   "add",
		Short: "Add a role to a group",
		Long: `Add a role to a ServiceNow group.

Examples:
  jsn grouproles add --group "IT Support" --role "itil"
  jsn gr add -g "IT Support" -r "admin"`,
		RunE: func(cmd *cobra.Command, args []string) error {
			if groupName == "" || roleName == "" {
				return fmt.Errorf("--group and --role are required")
			}

			app := appctx.FromContext(cmd.Context())
			ctx := cmd.Context()

			return addGroupRole(ctx, app, groupName, roleName)
		},
	}

	cmd.Flags().StringVarP(&groupName, "group", "g", "", "Group name (required)")
	cmd.Flags().StringVarP(&roleName, "role", "r", "", "Role name to add (required)")
	_ = cmd.MarkFlagRequired("group")
	_ = cmd.MarkFlagRequired("role")

	return cmd
}

func newGroupRolesRemoveCmd() *cobra.Command {
	var (
		groupName string
		roleName  string
	)

	cmd := &cobra.Command{
		Use:   "remove",
		Short: "Remove a role from a group",
		Long: `Remove a role from a ServiceNow group.

Examples:
  jsn grouproles remove --group "IT Support" --role "itil"
  jsn gr remove -g "IT Support" -r "admin"`,
		RunE: func(cmd *cobra.Command, args []string) error {
			if groupName == "" || roleName == "" {
				return fmt.Errorf("--group and --role are required")
			}

			app := appctx.FromContext(cmd.Context())
			ctx := cmd.Context()

			return removeGroupRole(ctx, app, groupName, roleName)
		},
	}

	cmd.Flags().StringVarP(&groupName, "group", "g", "", "Group name (required)")
	cmd.Flags().StringVarP(&roleName, "role", "r", "", "Role name to remove (required)")
	_ = cmd.MarkFlagRequired("group")
	_ = cmd.MarkFlagRequired("role")

	return cmd
}

func newGroupRolesListCmd() *cobra.Command {
	var groupName string

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List roles of a group",
		Long: `List all roles assigned to a ServiceNow group.

Examples:
  jsn grouproles list --group "IT Support"
  jsn gr list -g "IT Support"`,
		RunE: func(cmd *cobra.Command, args []string) error {
			if groupName == "" {
				return fmt.Errorf("--group is required")
			}

			app := appctx.FromContext(cmd.Context())
			ctx := cmd.Context()

			return listGroupRoles(ctx, app, groupName)
		},
	}

	cmd.Flags().StringVarP(&groupName, "group", "g", "", "Group name (required)")
	_ = cmd.MarkFlagRequired("group")

	return cmd
}

func addGroupRole(ctx context.Context, app *appctx.App, groupName, roleName string) error {
	// Find group sys_id
	groupSysID, err := findGroupSysID(ctx, app, groupName)
	if err != nil {
		return err
	}

	// Find role sys_id
	roleSysID, err := findRoleSysID(ctx, app, roleName)
	if err != nil {
		return err
	}

	// Check if role assignment already exists
	params := url.Values{}
	params.Set("sysparm_query", fmt.Sprintf("group=%s^role=%s", groupSysID, roleSysID))
	params.Set("sysparm_limit", "1")
	params.Set("sysparm_fields", "sys_id")

	existing, err := app.SDK.List(ctx, "sys_group_has_role", params)
	if err != nil {
		return fmt.Errorf("failed to check existing role assignment: %w", err)
	}
	if len(existing) > 0 {
		return fmt.Errorf("group %s already has role %s", groupName, roleName)
	}

	// Create the role assignment record
	recordData := map[string]any{
		"group": groupSysID,
		"role":  roleSysID,
	}

	record, err := app.SDK.Create(ctx, "sys_group_has_role", recordData)
	if err != nil {
		return fmt.Errorf("failed to add role to group: %w", err)
	}

	return app.OK(record,
		output.WithSummary(fmt.Sprintf("Added role %s to %s", roleName, groupName)),
		output.WithBreadcrumbs(
			output.Breadcrumb{
				Action:      "list",
				Cmd:         fmt.Sprintf("jsn grouproles list --group \"%s\"", groupName),
				Description: "List group roles",
			},
			output.Breadcrumb{
				Action:      "remove",
				Cmd:         fmt.Sprintf("jsn grouproles remove --group \"%s\" --role \"%s\"", groupName, roleName),
				Description: "Remove this role",
			},
		),
	)
}

func removeGroupRole(ctx context.Context, app *appctx.App, groupName, roleName string) error {
	// Find group sys_id
	groupSysID, err := findGroupSysID(ctx, app, groupName)
	if err != nil {
		return err
	}

	// Find role sys_id
	roleSysID, err := findRoleSysID(ctx, app, roleName)
	if err != nil {
		return err
	}

	// Find the role assignment record
	params := url.Values{}
	params.Set("sysparm_query", fmt.Sprintf("group=%s^role=%s", groupSysID, roleSysID))
	params.Set("sysparm_limit", "1")
	params.Set("sysparm_fields", "sys_id")

	records, err := app.SDK.List(ctx, "sys_group_has_role", params)
	if err != nil {
		return fmt.Errorf("failed to find role assignment: %w", err)
	}
	if len(records) == 0 {
		return fmt.Errorf("group %s does not have role %s", groupName, roleName)
	}

	assignmentSysID := getDisplayValue(records[0], "sys_id")

	// Delete the role assignment record
	if err := app.SDK.Delete(ctx, "sys_group_has_role", assignmentSysID); err != nil {
		return fmt.Errorf("failed to remove role from group: %w", err)
	}

	return app.OK(map[string]string{
		"role":  roleName,
		"group": groupName,
	},
		output.WithSummary(fmt.Sprintf("Removed role %s from %s", roleName, groupName)),
		output.WithBreadcrumbs(
			output.Breadcrumb{
				Action:      "list",
				Cmd:         fmt.Sprintf("jsn grouproles list --group \"%s\"", groupName),
				Description: "List group roles",
			},
			output.Breadcrumb{
				Action:      "add",
				Cmd:         fmt.Sprintf("jsn grouproles add --group \"%s\" --role \"%s\"", groupName, roleName),
				Description: "Add role back to group",
			},
		),
	)
}

func listGroupRoles(ctx context.Context, app *appctx.App, groupName string) error {
	// Find group sys_id
	groupSysID, err := findGroupSysID(ctx, app, groupName)
	if err != nil {
		return err
	}

	// List roles
	params := url.Values{}
	params.Set("sysparm_query", fmt.Sprintf("group=%s", groupSysID))
	params.Set("sysparm_display_value", "all")
	params.Set("sysparm_fields", "role")
	params.Set("sysparm_limit", "100")

	records, err := app.SDK.List(ctx, "sys_group_has_role", params)
	if err != nil {
		return fmt.Errorf("failed to list group roles: %w", err)
	}

	var roles []map[string]string
	for _, record := range records {
		if roleField, ok := record["role"].(map[string]any); ok {
			roleName := ""
			if dv, ok := roleField["display_value"].(string); ok {
				roleName = dv
			}
			roles = append(roles, map[string]string{
				"role": roleName,
			})
		}
	}

	return app.OK(map[string]any{
		"group": groupName,
		"count": len(roles),
		"roles": roles,
	},
		output.WithSummary(fmt.Sprintf("%d role(s) assigned to %s", len(roles), groupName)),
		output.WithBreadcrumbs(
			output.Breadcrumb{
				Action:      "add",
				Cmd:         fmt.Sprintf("jsn grouproles add --group \"%s\" --role \"<role_name>\"", groupName),
				Description: "Add a role to this group",
			},
			output.Breadcrumb{
				Action:      "group",
				Cmd:         fmt.Sprintf("jsn groups show \"%s\"", groupName),
				Description: "View group details",
			},
		),
	)
}

// findRoleSysID finds a role by name and returns its sys_id
func findRoleSysID(ctx context.Context, app *appctx.App, roleName string) (string, error) {
	params := url.Values{}
	params.Set("sysparm_query", "name="+roleName)
	params.Set("sysparm_limit", "1")
	params.Set("sysparm_fields", "sys_id")

	records, err := app.SDK.List(ctx, "sys_user_role", params)
	if err != nil {
		return "", fmt.Errorf("failed to find role: %w", err)
	}
	if len(records) == 0 {
		return "", fmt.Errorf("role not found: %s", roleName)
	}

	return getDisplayValue(records[0], "sys_id"), nil
}

File: src/commands/grouproles.js
import { formatRecordForDisplay, getStringField, resolveFieldsParam } from '../helpers.js';

export function groupRolesCmd(wrap) {
  return {
    command: 'grouproles [subcommand]',
    aliases: ['grouprole', 'grole', 'gr'],
    describe: 'Manage group roles',
    builder: (yargs) => {
      return yargs
        .command({
          command: 'list',
          aliases: ['ls'],
          describe: 'List group roles',
          builder: (y) => y
            .option('query', { type: 'string', describe: 'Encoded query (e.g. "nameLIKEincident" or "active=true")' })
            .option('columns', { alias: ['c', 'fields'], type: 'string', describe: 'Comma-separated columns (e.g. "number,short_description")' })
            .option('limit', { alias: 'l', type: 'number', default: 20, describe: 'Max records' }),
          handler: wrap(async (argv, app) => {
            const columns = argv.columns ? argv.columns.split(',') : ['group.name', 'role.name'];
            const params = new URLSearchParams();
            params.set('sysparm_limit', String(argv.limit));
            params.set('sysparm_display_value', 'all');
            const fields = resolveFieldsParam(columns);
            if (fields) params.set('sysparm_fields', fields);
            if (argv.query) params.set('sysparm_query', argv.query);
            const records = await app.sdk.list('sys_group_has_role', params);
            app.ok({
              table: 'sys_group_has_role',
              count: records.length,
              columns,
              records: fields ? records.map(r => formatRecordForDisplay(r, columns)) : records,
              context: { instance_url: app.getEffectiveInstance() },
            }, { summary: `${records.length} group role(s)` });
          }),
        })
        .command({
          command: 'add',
          describe: 'Add a role to a group',
          builder: (y) => y
            .option('group', { alias: 'g', type: 'string', demandOption: true, describe: 'Group name or sys_id (required)' })
            .option('role', { alias: 'r', type: 'string', demandOption: true, describe: 'Role name or sys_id (required)' }),
          handler: wrap(async (argv, app) => {
            const groupName = argv.group;
            const roleName = argv.role;

            // Resolve group sys_id
            const isGroupSysID = groupName.length === 32 && /^[0-9a-fA-F]+$/.test(groupName);
            const groupParams = new URLSearchParams();
            groupParams.set('sysparm_query', isGroupSysID ? `sys_id=${groupName}` : `name=${groupName}`);
            groupParams.set('sysparm_limit', '1');
            groupParams.set('sysparm_fields', 'sys_id');
            const groupRecords = await app.sdk.list('sys_user_group', groupParams);
            if (groupRecords.length === 0) {
              throw new Error(`Group not found: ${groupName}`);
            }
            const groupSysID = getStringField(groupRecords[0], 'sys_id');

            // Resolve role sys_id
            const isRoleSysID = roleName.length === 32 && /^[0-9a-fA-F]+$/.test(roleName);
            const roleParams = new URLSearchParams();
            roleParams.set('sysparm_query', isRoleSysID ? `sys_id=${roleName}` : `name=${roleName}`);
            roleParams.set('sysparm_limit', '1');
            roleParams.set('sysparm_fields', 'sys_id');
            const roleRecords = await app.sdk.list('sys_user_role', roleParams);
            if (roleRecords.length === 0) {
              throw new Error(`Role not found: ${roleName}`);
            }
            const roleSysID = getStringField(roleRecords[0], 'sys_id');

            // Check if role assignment already exists
            const checkParams = new URLSearchParams();
            checkParams.set('sysparm_query', `group=${groupSysID}^role=${roleSysID}`);
            checkParams.set('sysparm_limit', '1');
            checkParams.set('sysparm_fields', 'sys_id');
            const existing = await app.sdk.list('sys_group_has_role', checkParams);
            if (existing.length > 0) {
              throw new Error(`Role ${roleName} is already assigned to group ${groupName}`);
            }

            // Create the role assignment
            const record = await app.sdk.create('sys_group_has_role', {
              group: groupSysID,
              role: roleSysID,
            });
            app.ok(record, {
              summary: `Added role ${roleName} to ${groupName}`,
              breadcrumbs: [
                { action: 'list', cmd: `jsn grouproles list --query "group.name=${groupName}"`, description: 'List group roles' },
                { action: 'remove', cmd: `jsn grouproles remove --group "${groupName}" --role "${roleName}"`, description: 'Remove this role' },
              ],
            });
          }),
        })
        .command({
          command: 'remove',
          describe: 'Remove a role from a group',
          builder: (y) => y
            .option('group', { alias: 'g', type: 'string', demandOption: true, describe: 'Group name or sys_id (required)' })
            .option('role', { alias: 'r', type: 'string', demandOption: true, describe: 'Role name or sys_id (required)' }),
          handler: wrap(async (argv, app) => {
            const groupName = argv.group;
            const roleName = argv.role;

            // Resolve group sys_id
            const isGroupSysID = groupName.length === 32 && /^[0-9a-fA-F]+$/.test(groupName);
            const groupParams = new URLSearchParams();
            groupParams.set('sysparm_query', isGroupSysID ? `sys_id=${groupName}` : `name=${groupName}`);
            groupParams.set('sysparm_limit', '1');
            groupParams.set('sysparm_fields', 'sys_id');
            const groupRecords = await app.sdk.list('sys_user_group', groupParams);
            if (groupRecords.length === 0) {
              throw new Error(`Group not found: ${groupName}`);
            }
            const groupSysID = getStringField(groupRecords[0], 'sys_id');

            // Resolve role sys_id
            const isRoleSysID = roleName.length === 32 && /^[0-9a-fA-F]+$/.test(roleName);
            const roleParams = new URLSearchParams();
            roleParams.set('sysparm_query', isRoleSysID ? `sys_id=${roleName}` : `name=${roleName}`);
            roleParams.set('sysparm_limit', '1');
            roleParams.set('sysparm_fields', 'sys_id');
            const roleRecords = await app.sdk.list('sys_user_role', roleParams);
            if (roleRecords.length === 0) {
              throw new Error(`Role not found: ${roleName}`);
            }
            const roleSysID = getStringField(roleRecords[0], 'sys_id');

            // Find the role assignment record
            const findParams = new URLSearchParams();
            findParams.set('sysparm_query', `group=${groupSysID}^role=${roleSysID}`);
            findParams.set('sysparm_limit', '1');
            findParams.set('sysparm_fields', 'sys_id');
            const assignmentRecords = await app.sdk.list('sys_group_has_role', findParams);
            if (assignmentRecords.length === 0) {
              throw new Error(`Role ${roleName} is not assigned to group ${groupName}`);
            }
            const assignmentSysID = getStringField(assignmentRecords[0], 'sys_id');

            // Delete the role assignment
            await app.sdk.delete('sys_group_has_role', assignmentSysID);
            app.ok({ group: groupName, role: roleName }, {
              summary: `Removed role ${roleName} from ${groupName}`,
              breadcrumbs: [
                { action: 'list', cmd: `jsn grouproles list --query "group.name=${groupName}"`, description: 'List group roles' },
                { action: 'add', cmd: `jsn grouproles add --group "${groupName}" --role "${roleName}"`, description: 'Add role back to group' },
              ],
            });
          }),
        });
    },
    handler: () => {},
  };
}
